src/leap/bitmask/logs/safezmqhandler.py

```diff
@@ -0,0 +1,118 @@
+# -*- coding: utf-8 -*-
+# mainwindow.py
+# Copyright (C) 2013, 2014 LEAP
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+A thread-safe zmq handler for LogBook.
+"""
+import json
+import threading
+
+from logbook.queues import ZeroMQHandler
+from logbook import NOTSET
+
+import zmq
+
+
+class SafeZMQHandler(ZeroMQHandler):
+    """
+    A ZMQ log handler for LogBook that is thread-safe.
+
+    This log handler makes use of the existing zmq handler and if the user
+    tries to log something from a different thread than the one used to
+    create the handler a new socket is created for that thread.
+
+    Note: In ZMQ, Contexts are threadsafe objects, but Sockets are not.
+    """
+
+    def __init__(self, uri=None, level=NOTSET, filter=None, bubble=False,
+                 context=None, multi=False):
+
+        ZeroMQHandler.__init__(self, uri, level, filter, bubble, context,
+                               multi)
+
+        current_id = self._get_caller_id()
+        # we store the socket created on the parent
+        self._sockets = {current_id: self.socket}
```

Comment:
The current `SafeZMQHandler` uses the `ZeroMQHandler` constructor which creates a socket each time (https://github.com/mitsuhiko/logbook/blob/master/logbook/queues.py#L187).
The purpose of this attribute is to prevent cases in which we use the same logger in different threads. For instance when we (in the same file) `deferToThread` a method/function, we are using the same logger/socket without calling `get_logger` again.

If we want to reuse the socket, we need to rewrite the constructor instead of calling the `ZeroMQHandler`'s one. The best approach for this case maybe is to inherit directly from `logbook.Handler`.